crates/bin/pd/src/main.rs

```diff
@@ -134,102 +129,14 @@ async fn main() -> anyhow::Result<()> {
-                // Sets a timeout for all gRPC requests, but note that in the case of streaming
-                // requests, the timeout is only applied to the initial request. This means that
-                // this does not prevent long lived streams, for example to allow clients to obtain
-                // new blocks.
-                // .timeout(std::time::Duration::from_secs(7))
-                // Wrap each of the gRPC services in a tonic-web proxy:
-                .add_service(we(StorageQueryServiceServer::new(StorageServer::new(
-                    storage.clone(),
-                ))))
-                .add_service(we(AppQueryServiceServer::new(AppServer::new(
-                    storage.clone(),
-                ))))
-                .add_service(we(CompactBlockQueryServiceServer::new(
-                    CompactBlockServer::new(storage.clone()),
-                )))
-                .add_service(we(DexQueryServiceServer::new(DexServer::new(
-                    storage.clone(),
-                ))))
-                .add_service(we(FeeQueryServiceServer::new(FeeServer::new(
-                    storage.clone(),
-                ))))
-                .add_service(we(GovernanceQueryServiceServer::new(
-                    GovernanceServer::new(storage.clone()),
-                )))
-                .add_service(we(SctQueryServiceServer::new(SctServer::new(
-                    storage.clone(),
-                ))))
-                .add_service(we(ShieldedPoolQueryServiceServer::new(
-                    ShieldedPoolServer::new(storage.clone()),
-                )))
-                .add_service(we(StakeQueryServiceServer::new(StakeServer::new(
-                    storage.clone(),
-                ))))
-                .add_service(we(ClientQueryServer::new(ibc.clone())))
-                .add_service(we(ChannelQueryServer::new(ibc.clone())))
-                .add_service(we(ConnectionQueryServer::new(ibc.clone())))
-                .add_service(we(TendermintProxyServiceServer::new(tm_proxy.clone())))
-                .add_service(we(tonic_reflection::server::Builder::configure()
-                    .register_encoded_file_descriptor_set(penumbra_proto::FILE_DESCRIPTOR_SET)
-                    .build()
-                    .with_context(|| "could not configure grpc reflection service")?));
-
+            let mut grpc_server = penumbra_grpc_server::new(&storage)?.add_service(
+                tonic_web::enable(TendermintProxyServiceServer::new(tm_proxy.clone())),
+            );
             if enable_expensive_rpc {
-                grpc_server = grpc_server.add_service(we(SimulationServiceServer::new(
-                    DexServer::new(storage.clone()),
-                )));
+                use penumbra_dex::component::rpc::Server as DexServer;
+                grpc_server = grpc_server.add_service(tonic_web::enable(
+                    SimulationServiceServer::new(DexServer::new(storage.clone())),
+                ));
```

Comment:
for simplicity i left these both here. i don't have reason to believe yet that they will be needed by our tests, but they'd be easy to move over later if needed.

Reply:
https://github.com/penumbra-zone/penumbra/pull/3980/#issuecomment-1986436183

Reply:
https://github.com/penumbra-zone/penumbra/pull/3980#issuecomment-1986436183

Reply:
the order of `add_service` is load-bearing